Interactive test harness improvements

garethsb-sony opened the pull request:
* repeatable tests from the 'result' page, including ability to just re-run failed tests (though 'auto' is something of a challenge!)
* multi-select tests on the 'index' page
* changes to the underlying test runner to allow a list of test cases to be passed in

Resolves #150.

garethsb-sony commented:
> Looks like it does the job. A few minor issues/comments:

Thanks for the quick review, Andrew.

> * If you re-run some tests, the 'Go Back' button then results in form re-submission errors in the browser

Ok. Don't think we had noticed that, but I do wonder if Go Back just doing history - 1 is now what we want in any case? Maybe just a link back to the index page?

> If you select 'all' and one or more other tests from the selection box it runs some tests twice. I wonder if this just needs some validation / JS to prevent selecting 'all' with anything else?

Huh, we had tested that, and had it working exactly like that preventing anything else with 'all'. Will double check.

> Could the 'test selection' box be made a little taller and/or provide a handle on the bottom right to allow the user to make it bigger? I'm not 100% sure if this is possible with select boxes as well as plain test input ones.

Agreed. Will investigate.

> Checking 'Failed tests' on a results page with no failed tests seems to make it stick on. This doesn't cause a particular problem, but I wonder if the box could just be disabled if there are no failed tests?

Yeah, I shrugged when I encountered that, but disabling sounds reasonable.

> There are a couple of minor CSS things it might be nice to neaten up such as the NMOS Test heading at the top of the results page not quite being centred. Obviously no big deal.

I'm definitely CSS-challenged, yo help in this area much appreciated!

andrewbonney commented:
> Ok. Don't think we had noticed that, but I do wonder if Go Back just doing history - 1 is now what we want in any case? Maybe just a link back to the index page?

Yeah, I think this is how it worked before but we modified it prior to the completion of the session storage stuff. As long as the boxes still come back populated that should be fine.

> I'm definitely CSS-challenged, so help in this area much appreciated!

I'll take a look, but if I haven't had chance before the rest of the PR's done it can just be addressed after merge.

garethsb-sony commented:
> > > If you re-run some tests, the 'Go Back' button then results in form re-submission errors in the browser
> >
> > Ok. Don't think we had noticed that, but I do wonder if Go Back just doing history - 1 is now what we want in any case? Maybe just a link back to the index page?
> 
> Yeah, I think this is how it worked before but we modified it prior to the completion of the session storage stuff. As long as the boxes still come back populated that should be fine.

Done.

> > I'm definitely CSS-challenged, so help in this area much appreciated!
> 
> I'll take a look, but if I haven't had chance before the rest of the PR's done it can just be addressed after merge.

We've done a tiny bit of tidying up.

> > If you select 'all' and one or more other tests from the selection box it runs some tests twice. I wonder if this just needs some validation / JS to prevent selecting 'all' with anything else?
> 
> Huh, we had tested that, and had it working exactly like that preventing anything else with 'all'. Will double check.

Can't reproduce in Chrome 72, Firefox 65 or Edge 42. When you try to multi-select 'All' and anything else, the others are immediately deselected for us.

Could you give us some more info, Andrew?

> > Could the 'test selection' box be made a little taller and/or provide a handle on the bottom right to allow the user to make it bigger? I'm not 100% sure if this is possible with select boxes as well as plain test input ones.
> 
> Agreed. Will investigate.

Resizability added. Works in Chrome and Firefox, but not Edge.

> > Checking 'Failed tests' on a results page with no failed tests seems to make it stick on. This doesn't cause a particular problem, but I wonder if the box could just be disabled if there are no failed tests?
> 
> Yeah, I shrugged when I encountered that, but disabling sounds reasonable.

Fixed.
